utils/firestore.js

```diff
@@ -2305,6 +2305,59 @@ const getRestrictedFields = async () => {
     return snapshot.docs[0].data().restrictedFields;
 }
 
+// This is for managing received boxes in BPTL only.
+// receivedTimestamp format has year, month, date. Time is not specified. 'YYYY-MM-DDT00:00:00.000Z'. Ex: '2023-08-30T00:00:00.000Z'
+const getSpecimensByReceivedDate = async (receivedTimestamp) => {
+    const { extractCollectionIdsFromBoxes, processSpecimenCollections } = require('./shared');
+    try {
+        const boxes = await getBoxesByReceivedDate(receivedTimestamp);
+        const collectionIdArray = extractCollectionIdsFromBoxes(boxes);
+        if (collectionIdArray.length === 0) {
+            return [];
+        }
+
+        const specimenCollections = await getSpecimensByCollectionIds(collectionIdArray, true);
+        const specimenData = processSpecimenCollections(specimenCollections, receivedTimestamp);
+
+        return specimenData;
+    } catch (error) {
+        error.message = `Error fetching specimens by received date: ${error.message}`;
+        throw error;
+    }
+}
+
+// This is for managing received boxes in BPTL only.
+// receivedTimestamp format has year, month, date. Time is not specified. 'YYYY-MM-DDT00:00:00.000Z'. Ex: '2023-08-30T00:00:00.000Z'
+const getBoxesByReceivedDate = async (receivedTimestamp) => {
+    const snapshot = await db.collection('boxes').where('926457119', '==', receivedTimestamp).get();
+    return snapshot.docs.map(doc => doc.data());
+}
+
+// Get biospecimen docs from collectionIdsArray (conceptId: 820476880). Ex: ['CXA123456', 'CXA234567', 'CXA345678']
+// Max 30 disjunctions for 'in' queries: array length <= 30 for BPTL, <= 15 for site due to extra .where() clause in site query 
+const getSpecimensByCollectionIds = async (collectionIdsArray, isBPTL = false, siteCode) => {
+    try {
+        if (collectionIdsArray.length <= 30 && isBPTL) {
+            const snapshots = await db.collection('biospecimen').where('820476880', 'in', collectionIdsArray).get()
+            return snapshots.docs.map(doc => doc.data());
+        } else if (collectionIdsArray.length <= 15 && !isBPTL) {
+            const snapshots = await db.collection('biospecimen').where('820476880', 'in', collectionIdsArray).where('827220437', '==', siteCode).get();
+            return snapshots.docs.map(doc => doc.data());
+        } else {
+            const queries = collectionIdsArray.map(collectionId => {
+                return isBPTL
+                    ? db.collection('biospecimen').where('820476880', '==', collectionId).get()
+                    : db.collection('biospecimen').where('820476880', '==', collectionId).where('827220437', '==', siteCode).get();
+            });
+            return (await Promise.all(queries))
```

Comment:
Highly suggest using Promise.all for array length <= 30.  It's a convenient way to handle asynchronous operation.

Reply:
Agreed that Promise.all() is a great way to send concurrent queries.

I'm leaning towards the 'in' operator when it can be used because it sends a single query as an array instead of Promise.all(), which sends separate queries for each item. This should cut down on overall query data processing (speed experiment).

Reply:
The handlings of small (<=30 or 15) and large (>30 or 15) array sizes are inconsistent. I suggest checking BPTL to decide query size, then cut array to chunks of 15 or 30 for processing.

Reply:
Update: Warren's suggested pattern is better than the existing. Updated to chunking and using Promise.all() to handle chunked get() requests with 'in' queries.